dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs

```diff
@@ -169,30 +168,21 @@ internal KernelFunction(string name, string? pluginName, string description, IRe
             // Quick check for cancellation after logging about function start but before doing any real work.
             cancellationToken.ThrowIfCancellationRequested();
 
-            // Invoke pre-invocation filter. If it requests cancellation, throw.
-            var invokingContext = kernel.OnFunctionInvokingFilter(this, arguments);
-
-            if (invokingContext?.Cancel is true)
+            var invocationContext = await kernel.OnFunctionInvocationAsync(this, arguments, async (context) =>
             {
-                throw new OperationCanceledException("A function filter requested cancellation before function invocation.");
-            }
+                // Invoke the function.
+                functionResult = await this.InvokeCoreAsync(kernel, context.Arguments, cancellationToken).ConfigureAwait(false);
 
-            // Invoke the function.
-            functionResult = await this.InvokeCoreAsync(kernel, arguments, cancellationToken).ConfigureAwait(false);
+                // Update context with result.
+                context.Result = functionResult;
+            }).ConfigureAwait(false);
 
-            // Invoke the post-invocation filter. If it requests cancellation, throw.
-            var invokedContext = kernel.OnFunctionInvokedFilter(arguments, functionResult);
-
-            if (invokedContext is not null)
-            {
-                // Apply any changes from the function filters to final result.
-                functionResult = new FunctionResult(this, invokedContext.ResultValue, functionResult.Culture, invokedContext.Metadata ?? functionResult.Metadata);
-            }
-
-            if (invokedContext?.Cancel is true)
-            {
-                throw new OperationCanceledException("A function filter requested cancellation after function invocation.");
-            }
+            // Apply any changes from the function filters to final result.
+            functionResult = new FunctionResult(
+                this,
+                invocationContext.Result?.Value ?? functionResult?.Value,
+                functionResult?.Culture,
+                functionResult?.Metadata);
```

Comment:
Are we not allowing filters to modify the metadata and culture?

Reply:
Fixed.